# Ghosn Laboratory <img src="https://user-images.githubusercontent.com/50965273/136699995-7f58cf18-2580-4831-aea2-5fccd3c11005.jpg" align="right"  height=200/>
# SuPERR-Seq
These files are intended to support our manuscript titled "Comprehensive multi-omics single-cell data integration reveals greater heterogeneity in the human immune system".
This work can be found freely available at [bioRxiv](https://www.biorxiv.org/content/10.1101/2021.07.25.453651v2).

- The workflow of SuPERR-Seq is shown below:  
  ![Figure](workflow_github.png)

- The cellranger filtered GEX and ADT matrices can also be downloaded directly from our data directory [here](https://github.com/Ghosn-Lab/SuPERR-Seq/blob/main/data/raw_GEX%2BADT.RData) in order to fully reproduce the figures and results of our manuscript.

## Normalizing Antibody derived tags (ADT) data
ADT data was normalized using R package "**dsb**" (https://github.com/niaid/dsb). The background population in the cellranger raw feature-barcode matrix of each sample was determined separately after removing cell barcodes that expressed 0 ADT UMIs. The cell barcodes from the cellranger filtered feature-barcode were considered as real cells.

The DSB normalized ADT matrices that also contain the feature of Total VDJ UMIs can be found under the data directory [here](https://github.com/Ghosn-Lab/SuPERR-Seq/tree/main/data/DSB_normalized_ADT)
## Manual Gating of single-cell sequencing data
In the manuscript "Comprehensive multi-omics single-cell data integration reveals greater heterogeneity in the human immune system" we refer to Manual Gating using a customized strategy of biaxal plots, which was implemented in MATLAB.  
Following the code contained in **'Anti_seq_manual_gating_BM.m'** and **'Anti_seq_manual_gating_PBMC.m'**, we can easily reproduce the polygons used to produce the gates, and therefore reproduce the results of the manuscript.
- The major lineages of immune cells were gated using canonical surface protein markers and total UMIs of the VDJ filtered. Cell barcodes that expressed mutually exclusive surface proteins were considered doublets and have been removed from downstream analysis. 

The gating strategy of PBMC can be found under data/Major_Lineages [here](https://github.com/Ghosn-Lab/SuPERR-Seq/blob/main/data/Major%20lineages/PBMC_gating_strategy.png).

The gating strategy of BM can be also found in the same directory [here](https://github.com/Ghosn-Lab/SuPERR-Seq/blob/main/data/Major%20lineages/BM_gating_strategy.png)

A vector of the gated identities is located in the same directory and may be used to directly reproduce CMS scores and figures from the paper.

## Integration of GEX data
The standard Seurat integration workflow was utilized to generate the integrated matrix: [vignette](https://satijalab.org/seurat/articles/integration_introduction.html.)

## Sub-clustering and classification of each major lineages
After defining the major lineages of each sample and integrating the GEX data, the conventional GEX analysis was done for each major lineage separately. The sub-clusters were annotated based on DEG (Wilcoxon rank-sum test) analyses.

The scripts we used in the paper can be found [here](https://github.com/Ghosn-Lab/SuPERR/tree/main/Main_figure_codes)
## VDJ analysis
The VDJ analysis was done using cellranger vdj and custom codes, the manual for which is reproduced [here](https://github.com/Ghosn-Lab/SuPERR-Seq/blob/main/tools/VDJ_Analysis/VDJ_Analysis_Manual.pdf). Circos plots were generated by mode 3 in the custom codes. Specifically, we replaced the values of column "**isotype**" in **lineage_processed_data.csv** with cluster labels and ran mode 3 to generate the circos plot in Figure. 5E.

## CFS

In the manuscript "Comprehensive multi-omics single-cell data integration reveals greater heterogeneity in the human immune system" we refer to a Cell Fidelity Statistics  (**CFS**).

The CFS is a scoring metric we apply to evaluate whether cells in a single-cell RNA-seq experiment retain the same classification variable (e.g. cell type, function, state) between replicate dataset analyses.
CFS is presented here exactly as in our complementary work "Data Matrix Normalization and Merging Strategies Minimize Batch-specific Systemic Variation in scRNA-Seq Data" which is freely available at [bioRxiv](https://www.biorxiv.org/content/10.1101/2021.08.18.456898v1).
More details of CFS may be found [here](https://github.com/Ghosn-Lab/BatchNorm)

The identities used to generate CMS scores in R have been made available as a pre-computed .rds object available [here](https://github.com/Ghosn-Lab/SuPERR-Seq/tree/main/data/Major%20lineages). More details on generating CFS scores from these identities may be found in the "methods" section of our manuscript.

The source code that calculated CFS can be found [here](https://github.com/Ghosn-Lab/SuPERR/blob/main/Main_figure_codes/Ben_CFS.R).
You will need to
```
source("Ben_CFS.R")
getCFS(object=Seurat_object,ident.1=Cell_type,ident.2=cluster_labels)
```



